### DISCUSSION TITLE
How to achieve a centered grid layout with rows that are left-aligned?

### DISCUSSION THREAD

**Discussion by bonsairobo** [marc2332/freya#854] [Q&A] [answered]
I'm trying to implement a grid layout, but it ends up aligned to the top-left corner:

![20240907_19h49m16s_grim](https://github.com/user-attachments/assets/a24ee323-3c12-40a6-962d-d7d16da02010)

I would like it centered (horizontally) in the window instead.

I've tried wrapping the grid in another element like so:

```rust
        rect {
            width: "100%",
            height: "100%",

            direction: "vertical",
            cross_align: "center",

            ...my grid element...
        }
```

But this doesn't seem to do anything.

Here's the [full code](https://gist.github.com/bonsairobo/b03bdd13ae727816df9fef8a921ad27c).

**Comment by marc2332** [comment, marked as answer]
hey! You were very close.

You just need to remove the 100% width and 100% height from the second rect. This way the inner rect will not be expanded to its parent size and instead will be shrinked to its children size, and thus it will be centered.

🔴 
```rs
 rsx!(
        rect {
            width: "100%",
            height: "100%",

            direction: "vertical",
            cross_align: "center",

            rect {
                width: "100%",
                height: "100%",

                // Children are rows.
                direction: "vertical",
                main_align: "start",
                cross_align: "start",
```

✅ 
```rs
 rsx!(
        rect {
            width: "100%",
            height: "100%",

            direction: "vertical",
            cross_align: "center",

            rect {

                // Children are rows.
                direction: "vertical", // You can remove this btw, it is the default value
                main_align: "start",   // Same for this
                cross_align: "start",  // Same for this
```

![image](https://github.com/user-attachments/assets/d1f4a2d7-882e-4c35-8892-ba011f0cc51b)

**Reply by bonsairobo** [reply]
Excellent! Thanks so much.

**Reply by bonsairobo** [reply]
I actually have a followup question. If I wrap this grid in a `ScrollView`, then I lose the centered alignment. I've tried a few different parameters for `ScrollViewProps` but I can't seem to center it. Do you know how to make it work?

**Reply by marc2332** [reply]
hey again, I think I should make ScrollView somehow more flexible to make it shrink to its inner elements. For now could you try specifying a fixed width for now?

```rs
    rsx!(
        rect {
            width: "100%",
            height: "100%",

            direction: "vertical",
            cross_align: "center",

            ScrollView {
                theme: theme_with!(ScrollViewTheme {
                    width: "1500".into()
                }),
                for (n_row, row) in grid.read().rows.iter().enumerate() {
```

**Reply by marc2332** [reply]
I got some ideas but I need time to work on it, will update when I have something. https://github.com/marc2332/freya/issues/857

**Reply by bonsairobo** [reply]
OK thanks Marc. I'll subscribe to that issue and just leave a comment in my code for now.

**Reply by marc2332** [reply]
hey! Sorry, you may not be able to center a ScrollView yet but you can definitely center its content:

```rs
rsx!(
        ScrollView {
            rect {
                width: "100%",
                direction: "vertical",
                cross_align: "center",

                for (n_row, row) in grid.read().rows.iter().enumerate() {
```

**Reply by marc2332** [reply]
Just opened https://github.com/marc2332/freya/pull/858 ! You should now be able to specify a `width: auto` in the ScrollView:

```rs
 rsx!(
        rect {
            width: "100%",
            height: "100%",

            direction: "vertical",
            cross_align: "center",

            ScrollView {
                width: "auto",
                for (n_row, row) in grid.read().rows.iter().enumerate() {
```
